File: src/main/kotlin/rchabot/services/tournament/TournamentService.kt
package rchabot.services.tournament

import dev.forkhandles.result4k.Result4k
import org.bson.types.ObjectId
import rchabot.common.page.Page
import rchabot.common.page.PageRequest
import rchabot.services.player.bo.PlayerBO
import rchabot.services.tournament.bo.TournamentBO

interface TournamentService {

    suspend fun findAll(pageRequest: PageRequest): Page<TournamentBO>

    suspend fun create(name: String): Result4k<TournamentBO, Error>

    suspend fun findById(tournamentId: ObjectId): TournamentBO

    suspend fun delete(tournamentId: ObjectId)

    suspend fun getLeaderboard(tournamentId: ObjectId): List<PlayerBO>

    suspend fun registerPlayer(tournamentId: ObjectId, playerName: String): Result4k<List<PlayerBO>, Error>

    suspend fun updatePlayerScore(tournamentId: ObjectId, player: PlayerBO): List<PlayerBO>

    suspend fun findPlayer(tournamentId: ObjectId, playerName: String): PlayerBO

    suspend fun deletePlayers(tournamentId: ObjectId): TournamentBO
}

File: src/main/angular/tournament-webapp/src/app/shared/service/tournament.service.ts
import {Injectable} from '@angular/core';
import {HttpClient, HttpParams} from "@angular/common/http";
import {Observable} from "rxjs";
import {Tournament} from "../model/tournament.model";
import {Player} from "../model/player.model";
import {Page} from "../model/page.model";

@Injectable({
  providedIn: 'root'
})
export class TournamentService {

  baseUrl = "/api"

  constructor(private http: HttpClient) {
  }

  findAll(page: number, size: number): Observable<Page<Tournament>> {
    const params = new HttpParams()
      .set("page", page)
      .set("size", size)
    return this.http.get<Page<Tournament>>(`${this.baseUrl}/tournament`, {params: params})
  }

  find(id: string): Observable<Tournament> {
    return this.http.get<Tournament>(`${this.baseUrl}/tournament/${id}`)
  }

  getLeaderboard(id: string): Observable<Array<Player>> {
    return this.http.get<Array<Player>>(`${this.baseUrl}/tournament/${id}/leaderboard`)
  }

  create(name: string): Observable<Tournament> {
    const params = new HttpParams()
      .set("name", name)
    return this.http.post<Tournament>(
      `${this.baseUrl}/tournament`,
      {},
      {params: params}
    )
  }

  delete(tournamentId: string): Observable<any> {
    return this.http.delete(`${this.baseUrl}/tournament/${tournamentId}`)
  }

  registerPlayer(tournamentId: string, player: Player): Observable<Array<Player>> {
    const params = new HttpParams()
      .set("playerName", player.playerName)
    return this.http.post<Array<Player>>(
      `${this.baseUrl}/tournament/${tournamentId}/players`,
      {},
      {params: params}
    )
  }

  updatePlayerScore(tournamentId: string, player: Player): Observable<Array<Player>> {
    const params = new HttpParams()
      .set("score", player.score)
    return this.http.put<Array<Player>>(
      `${this.baseUrl}/tournament/${tournamentId}/players/${player.playerName}/scores`,
      {},
      {params: params}
    )
  }

  deletePlayers(tournamentId: string): Observable<Tournament> {
    return this.http.delete<Tournament>(`${this.baseUrl}/tournament/${tournamentId}/players`)
  }
}
